### PR TITLE
DD4Hep: DDCutTubsFromPoints Algorithm

### PR DESCRIPTION
**Pull request by vargasa**
DDCutTubsFromPoints Algorithm migration to `dd4hep`

Validation:

```
cmsRun DetectorDescription/DDCMS/test/python/testDDCutTubsFromPointsAlgorithm.py
./cmsShow -c simGeo.fwc --sim-geom-file cmsDD4HepGeom.root --tgeo-name=testCutTubsFromPointsAlgo
```

### CODE REVIEW

**Comment by ianna** [comment]
please test

**Comment by cvuosalo** [comment]
@cmsbuild please test

**Comment by ianna** [comment]
please test

**Comment by ianna** [comment]
please abort test

**Comment by ianna** [comment]
please test with #27584

**Comment by vargasa** [comment]
Running the test after rebasing:

Screenshot from 2019-07-24 11-44-44


```
./cmsShow -c simGeo.fwc --sim-geom-file cmsDD4HepGeom.root --tgeo-name=testCutTubsFromPointsAlgo
```

File: DetectorDescription/DDCMS/plugins/DDCutTubsFromPoints.cc [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+    s1 = s2;
+  }
+
+  assert(segments.size() >= 2);  // less would be special cases
+
+  dd4hep::Solid solid = segments[0];
+
+  // placement happens relative to the first member of the union
+  double shift = offsets[0];
+
```
Comment (cvuosalo) [inline DetectorDescription/DDCMS/plugins/DDCutTubsFromPoints.cc]: Isn't an assert for a non-empty `offsets` needed?
```
assert(!offsets.empty());
```

File: DetectorDescription/DDCMS/plugins/DDCutTubsFromPoints.cc [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+      double n_x_t = (P1_y_t * D2_z_t) - (D1_z_t * P2_y_t);
+      double n_y_t = (D1_z_t * P2_x_t) - (P1_x_t * D2_z_t);
+      double n_z_t = (P1_x_t * P2_y_t) - (P1_y_t * P2_x_t);
+
+      norm = sqrt(n_x_t * n_x_t + n_y_t * n_y_t + n_z_t * n_z_t);
+      n_x_t /= norm;
+      n_y_t /= norm;
+      n_z_t /= norm;
+
```
Comment (cvuosalo) [inline DetectorDescription/DDCMS/plugins/DDCutTubsFromPoints.cc]: A check or assert for `norm != 0.` should probably be added before dividing by `norm` here and after line 128.

Comment (vargasa) [reply DetectorDescription/DDCMS/plugins/DDCutTubsFromPoints.cc]: @cvuosalo As of now the main goal is to reproduce the same results with the migration. But additional checks do not hurt so I'll include them :+1:

**Comment by ianna** [comment]
please test

**Comment by ianna** [comment]
+1

**Comment by fabiocos** [comment]
+1

File: DetectorDescription/DDCMS/plugins/DDCutTubsFromPoints.cc [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    double z_t;  // +Z end (cuttubs t plane)
+    // radius is implicitly r_min
+  };
+
+  std::vector<Section> sections;
+
+  double r_min = args.value<double>("r_min");
+  double r_max = args.value<double>("r_max");
+  double z_pos = args.value<double>("z_pos");
+
```
Comment (vargasa) [inline DetectorDescription/DDCMS/plugins/DDCutTubsFromPoints.cc]: Fix: Original algorithm makes reference to: `rMin`,`rMax`, and `zPos` instead

File: DetectorDescription/DDCMS/data/testDDCutTubsFromPointsAlgorithm.xml [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+      <rSolid name="MotherOfAllBoxes"/>
+    </LogicalPart> 
+  </LogicalPartSection>
+  <Algorithm name="track:DDCutTubsFromPoints">
+    <rParent name="testDDCutTubsFromPointsAlgorithm:MotherOfAllBoxes"/>
+    <String name="SolidName" value="cutTubFromPoints"/>
+    <String name="Material" value="materials:Air"/>
+    <Numeric name="r_min" value="30.*cm"/>
+    <Numeric name="r_max" value="35.*cm"/>
+    <Numeric name="z_pos" value="10.*cm"/>
```
Comment (vargasa) [inline DetectorDescription/DDCMS/data/testDDCutTubsFromPointsAlgorithm.xml]: Not used in original algorithm

**Pull request merged**